### PR TITLE
Bipartite Subgraph + Subgraph refactoring.

### PR DESCRIPTION
**Pull request by wsad1**
This PR makes the following changes.
1. Added `bipartite_subgraph` to `utils` as discussed in #4001.
2. Refactored `subgraph`, by removing some if conditions.

### CODE REVIEW

File: test/utils/test_subgraph.py [hunk 1]
```diff
@@ -41,6 +42,27 @@ def test_subgraph():
         assert out[1].tolist() == [7, 8, 9, 10]
 
 
+def test_bipartite_subgraph():
+    edge_index = torch.tensor([[0, 5, 2, 3, 3, 4, 4, 3, 5, 5, 6],
+                               [0, 0, 3, 2, 0, 0, 2, 1, 2, 3, 1]])
+    edge_attr = torch.Tensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
+    idx = (torch.tensor([2, 3, 5], dtype=torch.long),
+           torch.tensor([2, 3], dtype=torch.long))
+    mask = (torch.tensor([0, 0, 1, 1, 0, 1, 0], dtype=torch.bool),
```
Comment (rusty1s) [inline test/utils/test_subgraph.py:51]: nit: Perhaps use `utils.index_to_mask` for the conversion.

File: test/utils/test_subgraph.py [hunk 2]
```diff
@@ -41,6 +42,27 @@ def test_subgraph():
...
+            torch.tensor([0, 0, 1, 1], dtype=torch.bool))
+    indices = ([2, 3, 5], [2, 3])
```
Comment (rusty1s) [inline test/utils/test_subgraph.py:53]: nit: Here as well? `indices = (idx[0].tolist(), idx[1].tolist())`

File: torch_geometric/utils/subgraph.py [hunk 3]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
         return edge_index, edge_attr
 
 
+def bipartite_subgraph(subset: Union[PairTensor, Tuple[List[int], List[int]]],
+                       edge_index: Tensor, edge_attr: Optional[Tensor] = None,
+                       relabel_nodes: bool = False,
+                       num_nodes: Tuple[int, int] = None,
+                       return_edge_mask: bool = False):
+    r"""Returns the induced subgraph of :obj:`(edge_index, edge_attr)`
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:79]:
```suggestion
    r"""Returns the induced subgraph of the bipartite graph :obj:`(edge_index, edge_attr)`
```

File: torch_geometric/utils/subgraph.py [hunk 4]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
...
+    containing the nodes in :obj:`subset`, for a bipartite graph.
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:80]:
```suggestion
    containing the nodes in :obj:`subset`.
```

File: torch_geometric/utils/subgraph.py [hunk 5]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
...
+
+    Args:
+        subset (PairTensor or tuple([int],[int])): The nodes to keep.
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:83]:
```suggestion
        subset (Tuple[Tensor, Tensor] or tuple([int],[int])): The nodes to keep.
```

File: torch_geometric/utils/subgraph.py [hunk 6]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
...
+        edge_index (LongTensor): The edge indices.
+        edge_attr (Tensor, optional): Edge weights or multi-dimensional
+            edge features. (default: :obj:`None`)
+        relabel_nodes (bool, optional): If set to :obj:`True`, the resulting
+            :obj:`edge_index` will be relabeled to hold consecutive indices
+            starting from zero. (default: :obj:`False`)
+        num_nodes (tuple, optional): The number of nodes.
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:90]:
```suggestion
        size (Tuple[int, int], optional): The size of the bipartite graph.
```
I think renaming that to `size` makes more sense.

File: torch_geometric/utils/subgraph.py [hunk 7]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
...
+        node_idx_j = torch.zeros(node_mask_j.size(0), dtype=torch.long,
+                                 device=device)
+        node_idx_i[subset[0]] = torch.arange(subset[0].sum().item(),
+                                             device=device)
+        node_idx_j[subset[1]] = torch.arange(subset[1].sum().item(),
+                                             device=device)
+        edge_index = torch.stack(
+            [node_idx_i[edge_index[0]], node_idx_j[edge_index[1]]])
+
+    if return_edge_mask:
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:131]: It would be great to test this :)

File: torch_geometric/utils/subgraph.py [hunk 8]
```diff
@@ -75,6 +71,69 @@ def subgraph(subset: Union[Tensor, List[int]], edge_index: Tensor,
...
+    if subset[0].dtype == torch.bool or subset[0].dtype == torch.uint8:
+        num_nodes = subset[0].size(0), subset[1].size(0)
+    else:
+        if num_nodes is None:
+            num_nodes = (maybe_num_nodes(edge_index[0]),
+                         maybe_num_nodes(edge_index[1]))
+        subset = (index_to_mask(subset[0], size=num_nodes[0]),
+                  index_to_mask(subset[1], size=num_nodes[1]))
+
+    node_mask_i, node_mask_j = subset[0], subset[1]
```
Comment (rusty1s) [inline torch_geometric/utils/subgraph.py:114]: Why do we need this here? Can't we simply do
```python
edge_mask = node_mask[0][edge_index[0]] & ...

**Review by rusty1s — APPROVED** [review]
LGTM. Thank you very much!

**Comment by rusty1s** [comment]
Please go ahead and merge if you are ready :)

**Comment by wsad1** [comment]
Thanks for your Review @rusty1s . Will merge in a few hours.

**Pull request merged**